File: dsp_modules/modules/overdrive/test/overdrive_test.c
#include <stdio.h>
#include "../../../test-tools/c/fopnchk/fopnchk.h"
#include <modules.h>

int main() {      
    printf("\n");
    printf("-------------\n");
    printf("-- TESTING --\n");
    printf("-------------\n");  

    float threshold = 0.3;
    int sampling_rate = 16000;
    int block_size = 128;

     // READ FILE
    FILE *f_in;
    f_in = fopen("../../../test-tools/signals/cleanguitar_sample.f32","rb");
    if (!f_in) {
        printf("Unable to open input file!");
        return -1;
    }  

    FILE *f_out = fopnchk("testvectors/c/output.f32","wb");

    if (!f_out) {
        printf("Unable to open output file!");
        return -1;
    }  
    
    t_overdrive_data overdrive_data;
    int init_code = overdrive_init(&overdrive_data, block_size,sampling_rate, threshold);
    if (init_code!=0) printf("Init error code: %i\n", init_code);

    float inbuf[block_size];
    while (fread(inbuf, sizeof(float), block_size, f_in) == block_size)
    {
        overdrive_apply(&overdrive_data, inbuf);

        fwrite(&overdrive_data.output, sizeof(float), block_size, f_out);
    }

    fopnchk_close_all();    
    return 0;
}

File: dsp_modules/modules/overdrive/test/overdrive_test.m
clear all
BLOCKSIZE = 128;
FS = 16000;

addpath('../matlab');

% run and write testvector to file
if ~exist('testvectors/matlab','dir')
  mkdir('testvectors/matlab');
end
fo0 = fopen('testvectors/matlab/output.f32','wb');


[in, fs0] = audioread('../../../test-tools/signals/cleanguitar_sample.wav');

threshold = 0.3;
o = overdrive(BLOCKSIZE, FS, threshold);

number_of_blocks = floor(length(in) / BLOCKSIZE);
output = [];

for i = 1:number_of_blocks   
    
    o.overdrive_apply(in(i*BLOCKSIZE - BLOCKSIZE + 1:i*BLOCKSIZE));
    output = [output; o.output_buffer];
   
end
% output = output./max(output);
% figure; plot(output);
% 
% maxyb = max(abs(output));
% output = output/maxyb;

audiowrite('overdrive.wav', output,FS)
fwrite(fo0, output, 'float32');
fclose('all');

